buffer/src/main/java/io/netty/buffer/PooledHeapByteBuf.java

```diff
@@ -245,4 +245,9 @@ public final long memoryAddress() {
     protected final ByteBuffer newInternalNioBuffer(byte[] memory) {
         return ByteBuffer.wrap(memory);
     }
+
+    @Override
+    protected void memoryClean() {
+        ByteBufUtil.setZero(memory, offset, length);
```

Comment:
I think you should set to zero `maxLength` bytes (applies to all `PooledByteBuf`s, probably also direct ones). After resizing to lower capacity there may be some data remaining in memory beyond current `length`